### PR TITLE
Add CANFD support to Linux CAN

### PR DESCRIPTION
**Pull request by alaa-az**
closes  https://github.com/ardera/flutter_packages/issues/24

### CODE REVIEW

**Comment by alaa-az** [comment]
@ardera  Can you please review this.

**Comment by ardera** [comment]
Hey @alaa-az sorry for the wait.

What CAN hardware did you test this with?

**Comment by alaa-az** [comment]
Hello @ardera on this: https://www.peak-system.com/PCAN-USB-FD.365.0.html?&L=1, also tested on virtual CAN

File: packages/linux_can/lib/src/can_device.dart [hunk 1]
```diff
@@ -311,6 +315,7 @@ class CanSocket implements Sink<CanFrame> {
   /// happen when sending lots of frames in a short time period. If [block] is false, this will throw a [LinuxError]
   /// with errno [EWOULDBLOCK] (value 22) in this case.
   Future<void> send(CanFrame frame, {bool block = true}) async {
+    assert(isFlexibleDataRate || frame is! CanFdFrame || frame is! CanFdFrameExtended);
```
Comment (ardera) [inline packages/linux_can/lib/src/can_device.dart:318]: I would rather throw an exception here, in my mind asserts are just to verify internal invariants that should always hold. Some user of this API forgetting to check `isFlexibleDataRate` and sending a wrong frame can easily happen

```suggestion
    if (!isFlexibleDataRate && frame is CanFdFrame) {
      throw ArgumentError.value(frame, 'frame', 'CAN controller does not support CAN FD.');
    }
```

File: packages/linux_can/lib/src/data_classes.dart [hunk 2]
```diff
@@ -566,6 +597,86 @@ class CanExtendedRemoteFrame extends CanFrame implements CanExtendedFrame, CanRe
...
+
+  @override
+  final CanFrameType type = CanFrameType.data;
+
+  @override
+  int get hashCode => Object.hash(id, data);
+
+  @override
+  bool operator ==(Object other) {
+    return other is CanFdFrameExtended && id == other.id && listsEqual(data, other.data);
```
Comment (ardera) [inline packages/linux_can/lib/src/data_classes.dart:673]:
```suggestion
    return other is CanFdFrameExtended && id == other.id && flags == other.flags && listsEqual(data, other.data);
```

File: packages/linux_can/lib/src/data_classes.dart [hunk 3]
```diff
@@ -566,6 +597,86 @@ class CanExtendedRemoteFrame extends CanFrame implements CanExtendedFrame, CanRe
...
+  final int flags;
+
+  @override
+  final CanFrameFormat format = CanFrameFormat.extended;
+
+  @override
+  final CanFrameType type = CanFrameType.data;
+
+  @override
+  int get hashCode => Object.hash(id, data);
```
Comment (ardera) [inline packages/linux_can/lib/src/data_classes.dart:669]:
```suggestion
  int get hashCode => Object.hash(id, flags, data);
```

File: packages/linux_can/lib/src/data_classes.dart [hunk 4]
```diff
@@ -566,6 +597,86 @@ class CanExtendedRemoteFrame extends CanFrame implements CanExtendedFrame, CanRe
...
+
+  @override
+  bool operator ==(Object other) {
+    return other is CanFdFrameExtended && id == other.id && listsEqual(data, other.data);
+  }
+
+  @override
+  String toString() => 'CanFdFrameExtended(id: $id, data: $data)';
```
Comment (ardera) [inline packages/linux_can/lib/src/data_classes.dart:677]:
```suggestion
  String toString() => 'CanFdFrameExtended(id: $id, flags: $flags, data: $data)';
```

File: packages/linux_can/lib/src/data_classes.dart [hunk 5]
```diff
@@ -566,6 +597,86 @@ class CanExtendedRemoteFrame extends CanFrame implements CanExtendedFrame, CanRe
...
+  final int flags;
+
+  @override
+  final CanFrameFormat format = CanFrameFormat.base;
+
+  @override
+  final CanFrameType type = CanFrameType.data;
+
+  @override
+  int get hashCode => Object.hash(id, data);
```
Comment (ardera) [inline packages/linux_can/lib/src/data_classes.dart:629]:
```suggestion
  int get hashCode => Object.hash(id, flags, data);
```

File: packages/linux_can/lib/src/data_classes.dart [hunk 6]
```diff
@@ -566,6 +597,86 @@ class CanExtendedRemoteFrame extends CanFrame implements CanExtendedFrame, CanRe
...
+
+  @override
+  bool operator ==(Object other) {
+    return other is CanFdFrame && id == other.id && listsEqual(data, other.data);
```
Comment (ardera) [inline packages/linux_can/lib/src/data_classes.dart:633]:
```suggestion
    return other is CanFdFrame && id == other.id && flags == other.flags && listsEqual(data, other.data);
```

File: packages/linux_can/lib/src/data_classes.dart [hunk 7]
```diff
@@ -566,6 +597,86 @@ class CanExtendedRemoteFrame extends CanFrame implements CanExtendedFrame, CanRe
...
+  }
+
+  @override
+  String toString() => 'CanFdFrame(id: $id, data: $data)';
```
Comment (ardera) [inline packages/linux_can/lib/src/data_classes.dart:637]:
```suggestion
  String toString() => 'CanFdFrame(id: $id, flags: $flags, data: $data)';
```

**Review by ardera — COMMENTED** [review]
Overall, looks good! Just some minor stuff. Thanks a lot for the contribution!

**Comment by alaa-az** [comment]
Fixes are done, Thank you in advance

File: packages/linux_can/lib/src/platform_interface.dart [hunk 8]
```diff
@@ -481,56 +480,36 @@ class PlatformInterface {
...
+        native.ref.data[index] = byte;
       }
 
       /// TODO: use sendmsg and MSG_DONTWAIT for non-blocking sends, and make blocking sends block in the kernel
       ///  instead of spinning here.
 
       // send the CAN frame.
       final ok = _retry(
-        () => libc.write(fd, ref, mtu),
+        () => libc.write(fd, native as ffi.Pointer<ffi.Void>, mtu),
```
Comment (ardera) [inline packages/linux_can/lib/src/platform_interface.dart:512]:
```suggestion
        () => libc.write(fd, native.cast<ffi.Void>(), mtu),
```

**Comment by ardera** [comment]
Sorry I pointed you in the wrong direction before, after this is fixed it's ready to merge :)

**Comment by alaa-az** [comment]
Don't worry I should have figured this out by myself.

**Comment by alaa-az** [comment]
Perfect solution, it works with Virtual CAN.

File: packages/linux_can/lib/src/data_classes.dart [hunk 9]
```diff
@@ -566,6 +588,87 @@ class CanExtendedRemoteFrame extends CanFrame implements CanExtendedFrame, CanRe
   String toString() => 'CanExtendedRemoteFrame(id: $id)';
 }
 
+class CanFdBaseFrame extends CanFrame implements CanBaseFrame, CanDataFrame, CanFdFrame {
+  const CanFdBaseFrame({required this.id, required this.data, required this.flags})
+      : assert(0 <= data.length &&
```
Comment (ardera) [inline packages/linux_can/lib/src/data_classes.dart:593]: just one super minor thing I just saw :)

```suggestion
      : assert((id & ~CAN_EFF_MASK == 0) && 0 <= data.length &&
```

**Comment by ardera** [comment]
merged it

thanks for the contribution!

**Comment by alaa-az** [comment]
Thank you for this repo

**Pull request merged**